### ISSUE TITLE
Implement full MobileNetV1 backward pass (~2000 lines)

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#3296]
The backward pass in examples/mobilenetv1-cifar10/train.mojo is a documented placeholder (issue #3187). Requires: 13 depthwise separable block backward passes (pointwise conv backward, BN backward, ReLU backward, depthwise conv backward), and SGD+momentum parameter updates. The depthwise backward is non-standard and requires per-channel gradient computation.

_Follow-up from #3084_

**Comment by mvillmow** [comment]
# Implementation Plan

I have all the context needed. Here is the implementation plan:

---

# Implementation Plan: Issue #3296 — MobileNetV1 Full Backward Pass

## 1. Objective

Replace the placeholder backward pass in `examples/mobilenetv1-cifar10/train.mojo` with a complete manual implementation covering:

- Forward pass with activation caching for all 13 depthwise separable blocks, the initial conv, and the final FC layer
- Full backward pass (in reverse layer order): FC → global avgpool → blocks 13→1 → initial conv
- Per-block backward: pointwise BN backward → pointwise conv backward → depthwise ReLU backward → depthwise BN backward → depthwise conv backward
- SGD with momentum parameter updates for all ~156 parameters

---

## 2. Approach

### Key Decisions

- **Pattern source**: Follow `examples/alexnet-cifar10/train.mojo` exactly — extract a `compute_gradients()` function that does forward + backward + update in one call. Do not use closures (Mojo limitation: can't capture non-copyable structs).
- **Depthwise backward**: `depthwise_conv2d_backward()` is already in `shared/core/conv.mojo:1057` and exported from `shared.core`. Use it directly — no custom implementation needed.
- **BN backward**: `batch_norm2d_backward()` is at `shared/core/normalization.mojo:463`. It requires the original pre-BN input `x` to be cached during forward pass.
- **Momentum velocities**: Maintain a `List[ExTensor]` of 156 velocities (6 initial + 13 blocks × 12 + 2 FC). Pass as `mut velocities` to each epoch call.
- **`update_parameters` method**: Add to `model.mojo:MobileNetV1` following the AlexNet pattern — accepts all gradients and velocities, calls `sgd_momentum_update_inplace`.
- **Heap corruption mitigation**: Keep backward pass tests ≤10 per test file (ADR-009). Cache activations use `._data.bitcast[Float32]()` — no cross-dtype casts.
- **Test count budget**: The existing `test_mobilenetv1_layers.mojo` already has 19 tests. New backward pass tests go into a separate file `test_mobilenetv1_backward.mojo` capped at 10 tests.

---

## 3. Files to Modify

### `examples/mobilenetv1-cifar10/train.mojo`

**Changes**:

- Replace `train_epoch()` with manual batch loop (no closure) — same as AlexNet
- Add `compute_gradients()` function with:
  - **Forward pass + caching** (~400 lines): Cache 13 × 6 intermediate tensors (dw_pre_bn, dw_bn_out, dw_relu_out, pw_pre_bn, pw_bn_out, pw_relu_out) plus initial conv, gap output, flattened
  - **Backward pass** (~1200 lines): Reverse through FC → gap → block 13 → … → block 1 → initial conv
  - **Parameter update** (~200 lines): SGD with momentum via `model.update_parameters()`
- Update `main()` to initialize velocity list and pass to training loop
- Remove the placeholder docstring comment

**Forward pass caching pattern** (per block):

```mojo
# Block N forward with caching
var blk_dw_pre_bn = depthwise_conv2d(prev_out, model.ds_block_N.dw_weights, model.ds_block_N.dw_bias, stride=S, padding=1)
var blk_dw_bn_out, _, _ = batch_norm2d(blk_dw_pre_bn, model.ds_block_N.dw_bn_gamma, model.ds_block_N.dw_bn_beta, model.ds_block_N.dw_bn_running_mean, model.ds_block_N.dw_bn_running_var, training)
var blk_dw_relu_out = relu(blk_dw_bn_out)
var blk_pw_pre_bn = conv2d(blk_dw_relu_out, model.ds_block_N.pw_weights, model.ds_block_N.pw_bias, stride=1, padding=0)
var blk_pw_bn_out, _, _ = batch_norm2d(blk_pw_pre_bn, model.ds_block_N.pw_bn_gamma, model.ds_block_N.pw_bn_beta, model.ds_block_N.pw_bn_running_mean, model.ds_block_N.pw_bn_running_var, training)
var blk_pw_relu_out = relu(blk_pw_bn_out)
```

**Backward pass pattern** (per block, in reverse):

```mojo
# Block N backward
var grad_blk_pw_bn_out = relu_backward(grad_blk_pw_relu_out, blk_pw_bn_out)
var grad_blk_pw_pre_bn, grad_blk_pw_bn_gamma, grad_blk_pw_bn_beta = batch_norm2d_backward(grad_blk_pw_bn_out, blk_pw_pre_bn, model.ds_block_N.pw_bn_gamma, model.ds_block_N.pw_bn_running_mean, model.ds_block_N.pw_bn_running_var, training)
var pw_grads = conv2d_backward(grad_blk_pw_pre_bn, blk_dw_relu_out, model.ds_block_N.pw_weights, stride=1, padding=0)
var grad_blk_dw_relu_out = pw_grads.grad_input
var grad_blk_dw_bn_out = relu_backward(grad_blk_dw_relu_out, blk_dw_bn_out)
var grad_blk_dw_pre_bn, grad_blk_dw_bn_gamma, grad_blk_dw_bn_beta = batch_norm2d_backward(grad_blk_dw_bn_out, blk_dw_pre_bn, model.ds_block_N.dw_bn_gamma, model.ds_block_N.dw_bn_running_mean, model.ds_block_N.dw_bn_running_var, training)
var dw_grads = depthwise_conv2d_backward(grad_blk_dw_pre_bn, prev_block_output, model.ds_block_N.dw_weights, stride=S, padding=1)
```

### `examples/mobilenetv1-cifar10/model.mojo`

**Changes**:

- Add `update_parameters(mut self, lr, momentum, <all 156 gradient tensors>, <all 156 velocity tensors>)` method to `MobileNetV1`
- Add import for `sgd_momentum_update_inplace` from `shared.training.optimizers`
- The method calls `sgd_momentum_update_inplace` for each of the 156 parameter/velocity pairs

---

## 4. Files to Create

### `tests/models/test_mobilenetv1_backward.mojo` (new)

New layerwise backward pass test file covering ≤10 tests:

| # | Test | Purpose |
|---|------|---------|
| 1 | `test_dw_block_backward_shapes()` | Verify grad shapes through one DS block backward |
| 2 | `test_dw_block_backward_values()` | Verify non-zero gradients flow through (training=True) |
| 3 | `test_pw_bn_backward_chain()` | Test BN backward → pointwise conv backward chain |
| 4 | `test_dw_bn_backward_chain()` | Test BN backward → depthwise conv backward chain |
| 5 | `test_initial_conv_backward()` | Initial conv backward shape verification |
| 6 | `test_global_avgpool_backward_shape()` | Already in layers file; verify integration |
| 7 | `test_sgd_momentum_update()` | Verify parameter changes after one SGD step |
| 8 | `test_full_block_backward_32to64()` | End-to-end backward through block 1 (32→64) |
| 9 | `test_full_block_backward_stride2()` | End-to-end backward through stride-2 block |
| 10 | `test_grad_nonzero_after_backward()` | Sanity check: all parameter grads non-zero |

---

## 5. Implementation Order

1. **Read** `shared/training/optimizers/sgd.mojo:161` to confirm `sgd_momentum_update_inplace` signature
2. **Add** `update_parameters()` to `model.mojo:MobileNetV1` — handles all 156 param/velocity pairs, calls `sgd_momentum_update_inplace` per parameter
3. **Create** `test_mobilenetv1_backward.mojo` with ≤10 tests — run against existing shared ops to confirm backward functions work correctly before wiring into train
4. **Rewrite** `train.mojo:train_epoch()` → manual batch loop (remove closure)
5. **Add** `compute_gradients()` to `train.mojo`:
   - Forward pass section with all cached variables
   - Backward pass section in reverse order
   - Delegate to `model.update_parameters()`
6. **Update** `main()` in `train.mojo`:
   - Initialize `velocities: List[ExTensor]` with 156 zero tensors matching each parameter shape
   - Pass velocities to `train_epoch` (and then to `compute_gradients`)
7. **Run** `just pre-commit-all` to check formatting
8. **Run** `just test-group "tests/models" "test_mobilenetv1*"` to verify all tests pass
9. **Create PR** linked to issue #3296

---

## 6. Verification

### Local Testing

```bash
# Run only mobilenetv1 test groups (matches CI)
just test-group "tests/models" "test_mobilenetv1_layers*"
just test-group "tests/models" "test_mobilenetv1_backward*"

# Verify formatting before commit
just pre-commit-all
```

### Correctness Checks

- All gradient tensors have same shape as their corresponding parameters
- Loss decreases over multiple batches (sanity test in `compute_gradients` — print first 3 batch losses)
- No heap corruption: test file stays ≤10 tests; use `._get_float32()` / `bitcast[Float32]()` only, no `_get_float64()` on Float32 data

### CI Validation

- `just test-group "tests/models" "test_mobilenetv1*"` must pass in <90s
- `just build` must succeed with zero warnings

---

## 7. Momentum Velocity Initialization

156 zero tensors in `main()` matching this order:

```text
# Initial conv block (6 tensors):
initial_conv_weights [32,3,3,3], initial_conv_bias [32],
initial_bn_gamma [32], initial_bn_beta [32],
initial_bn_running_mean [32], initial_bn_running_var [32]

# Per DS block (12 tensors × 13 blocks = 156 tensors):
dw_weights [C_in,1,3,3], dw_bias [C_in],
dw_bn_gamma [C_in], dw_bn_beta [C_in],
dw_bn_running_mean [C_in], dw_bn_running_var [C_in],
pw_weights [C_out,C_in,1,1], pw_bias [C_out],
pw_bn_gamma [C_out], pw_bn_beta [C_out],
pw_bn_running_mean [C_out], pw_bn_running_var [C_out]

# FC (2 tensors):
fc_weights [10,1024], fc_bias [10]
```

Total: 6 + 13×12 + 2 = **164 velocity tensors**

---

## 8. Critical Constraints

- **No closures** capturing `MobileNetV1` — Mojo can't capture non-copyable struct in closure. Use manual batch loop like AlexNet's `train_epoch`.
- **BN backward requires pre-BN input** — cache `blk_dw_pre_bn` and `blk_pw_pre_bn` (not post-BN) for `batch_norm2d_backward`.
- **`^` ownership transfer** when passing grad tensors to `update_parameters` — avoids copies of large tensors.
- **Stride-aware gradient flow** — for blocks with stride=2, `depthwise_conv2d_backward` handles the upsampling automatically (it uses the same stride as forward).
- **BN running stats not updated by backward** — `batch_norm2d_backward` returns `grad_input, grad_gamma, grad_beta`. The running_mean/var are only updated during forward pass.

---

## 9. Skills Used

### Skills Invoked During Planning

- `skills-registry-commands:advise` — Searched team knowledge base before starting

### Team Knowledge Base Skills Referenced (Prior Learnings)

- `fix-placeholder-code-ci` — Pattern for replacing placeholder backward stubs; warning about ALL dependent code needing to be commented, not just the stub call
- `investigate-mojo-heap-corruption` — Heap corruption at >13 tests; use native `bitcast[Float32]()`, no `_get_float64()` on Float32 data
- `fix-ci-mojo-test-failure` — Test count limits per file (≤10); use `just test-group` not `just test`
- `mojo-memory-check` — `^` ownership transfer for returned `List`/gradient tensors
- `mojo-simd-optimize` — Per-channel depthwise grad computation is SIMD candidate (already handled by `depthwise_conv2d_backward`)
- `train-model` — Training loop pattern reference (manual batch loop, not closure)

---
*Generated by Claude Code Planner*

**Comment by mvillmow** [comment]
Duplicate of #3181/#3184/#3187 — closing as part of bulk dedup
